### PR TITLE
✨ feat: add selectable types to Calendar `onSelect` 

### PR DESCRIPTION
**Pull request by hbatalhaStch**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

https://github.com/ant-design/ant-design/issues/40407

### 💡 Background and solution

This PR was already reviewed by @yoyo837 [here](https://github.com/ant-design/ant-design/pull/42075).

When I need to go to a specific date in a year and month different from the current selected date, it will require me to select the date 3 times just to get there. First I have to select the year, then select the month and only then select the day which is exhaustive and unnecessary IMO as it gives a really bad UX.
The issue is that the onSelect callback is called whenever we change the month or year. I propose an api that let you choose when you want the onSelect callback to be called, if it is when we change the year, when we change the month or only when change click on a day. We can also combine them.

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |      Make calendar selectable by `date`, `month` and `year`     |
| 🇨🇳 Chinese |      使日历可以按`date`、`month`和`year`进行选择      |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

**Comment by hbatalhaStch** [comment]
@MadCcc Hello, I am kindly requesting a review for this new feature. Thank you!

**Comment by yoyo837** [comment]
Sorry, most members are on holiday from April 29th to May 3rd.

**Comment by bombillazo** [comment]
@yoyo837 
any update on this?

File: components/calendar/index.zh-CN.md [hunk 1]
```diff
@@ -60,6 +60,7 @@ coverDark: https://mdn.alipayobjects.com/huamei_7uahnr/afts/img/A*-p-wQLik200AAA
 | value | 展示日期 | [dayjs](https://day.js.org/) | - |  |
 | onChange | 日期变化回调 | function(date: Dayjs) | - |  |
 | onPanelChange | 日期面板变化回调 | function(date: Dayjs, mode: string) | - |  |
+| selectable | 使日历可以按`date`、`month`和`year`进行选择 | ('date' \| 'year' \| 'month')[] | - |  |
```
Comment (afc163) [inline components/calendar/index.zh-CN.md:63]:
```suggestion
| selectable | 使日历可以按 `date`、`month` 和 `year` 进行选择 | ('date' \| 'year' \| 'month')[] | - |  |
```

File: components/calendar/Header.tsx [hunk 2]
```diff
@@ -147,7 +147,7 @@ export interface CalendarHeaderProps<DateType> {
   locale: Locale;
   mode: CalendarMode;
   fullscreen: boolean;
-  onChange: (date: DateType) => void;
+  onChange: (date: DateType, selectType: CalendarSelectable) => void;
```
Comment (afc163) [inline components/calendar/Header.tsx:150]:
```suggestion
  onChange: (date: DateType, selectInfo: {
    triggerType: CalendarSelectable,
  }) => void;
```

1. 放一个对象里，避免以后还有新的参数要加。
2. 改成 triggerType，日期更新不一定是用户选择行为。

File: components/calendar/generateCalendar.tsx [hunk 3]
```diff
@@ -332,7 +340,7 @@ function generateCalendar<DateType>(generateConfig: GenerateConfig<DateType>) {
           locale={contextLocale?.lang}
           generateConfig={generateConfig}
           cellRender={mergedCellRender}
-          onSelect={onInternalSelect}
+          onSelect={(date) => onInternalSelect(date, panelMode === 'date' ? 'date' : 'month')}
```
Comment (afc163) [inline components/calendar/generateCalendar.tsx:343]: onSelect 文档也需要更新？

File: components/calendar/generateCalendar.tsx [hunk 4]
```diff
@@ -198,9 +201,14 @@ function generateCalendar<DateType>(generateConfig: GenerateConfig<DateType>) {
...
 
-    const onInternalSelect = (date: DateType) => {
+    const onInternalSelect = (date: DateType, selectType?: CalendarSelectable) => {
       triggerChange(date);
 
+      if (selectable && selectable.length > 0 && selectType) {
+        if (!selectable.includes(selectType)) {
+          return;
+        }
+      }
```
Comment (afc163) [inline components/calendar/generateCalendar.tsx:211]:
```suggestion
      if (selectType && !(selectable || []).includes(selectType)) {
        return;
      }
```

File: components/calendar/__tests__/index.test.tsx [hunk 5]
```diff
@@ -86,6 +86,266 @@ describe('Calendar', () => {
...
+    onSelect.mockReset();
+
+    wrapper.rerender(<Calendar onSelect={onSelect} mode='month' selectable={['month', 'year']} />);
+
+    openSelect(wrapper.container, '.ant-picker-calendar-month-select');
+    fireEvent.click(
+      Array.from(wrapper.container.querySelectorAll('.ant-select-item-option')).at(5)!,
+    );
+    expect(onSelect.mock.calls.length).toBe(1);
+  });
```
Comment (afc163) [inline components/calendar/__tests__/index.test.tsx:347]: onSelect 感觉可以单拆一个 select.test.ts 了

File: components/calendar/generateCalendar.tsx (hunk 3 above)
Comment (hbatalhaStch) [reply components/calendar/generateCalendar.tsx:343]: No, I don't believe it is necessary to update the onSelect documentation.

File: components/calendar/__tests__/index.test.tsx [hunk 6]
```diff
@@ -86,6 +86,266 @@ describe('Calendar', () => {
...
+
```
Comment (afc163) [inline components/calendar/__tests__/index.test.tsx:348]: Should we add test case about `onChange`?

File: components/calendar/__tests__/index.test.tsx (hunk 5 above)
Comment (hbatalhaStch) [reply components/calendar/__tests__/index.test.tsx:347]: yes it does feel like it could be put in a separate  select.test.ts file, but is it necessary?

Comment (afc163) [reply components/calendar/__tests__/index.test.tsx:347]: It is long test file now, we could split it.

**Comment by zombieJ** [comment]
`selectable` is not a good design API. This mix the business logic into component. It should be like `onChange` update:

```tsx
interface Info {
  source: 'date' | 'month' | 'year';
}

type OnChange<T> = (date: T, info: Info);
type OnSelect<T> = (date: T, info: Info);
```

Developer can handle this by the logic with:
```tsx
const onMySelect = (date, { source }) => {
  if (source !== 'date') return;
};
```

**Comment by hbatalhaStch** [comment]
> `selectable` is not a good design API. This mix the business logic into component. It should be like `onChange` update:
> 
> ```tsx
> interface Info {
>   source: 'date' | 'month' | 'year';
> }
> 
> type OnChange<T> = (date: T, info: Info);
> type OnSelect<T> = (date: T, info: Info);
> ```
> 
> Developer can handle this by the logic with:
> 
> ```tsx
> const onMySelect = (date, { source }) => {
>   if (source !== 'date') return;
> };
> ```

Ok, I will do that

> `selectable` is not a good design API. This mix the business logic into component. It should be like `onChange` update:
> 
> ```tsx
> interface Info {
>   source: 'date' | 'month' | 'year';
> }
> 
> type OnChange<T> = (date: T, info: Info);
> type OnSelect<T> = (date: T, info: Info);
> ```
> 
> Developer can handle this by the logic with:
> 
> ```tsx
> const onMySelect = (date, { source }) => {
>   if (source !== 'date') return;
> };
> ```

So this [function](https://github.com/hbatalhaStch/ant-design/blob/ca1dd39dc0d92a8f7be9f5d62cbc4ed51cf08597/components/calendar/generateCalendar.tsx#L204) will be like this:
```ts
const triggerChange = (date: DateType, info: Info) => {
  setMergedValue(date);

  if (!isSameDate(date, mergedValue)) {
    // Trigger when month panel switch month
    if (
      (panelMode === 'date' && !isSameMonth(date, mergedValue)) ||
      (panelMode === 'month' && !isSameYear(date, mergedValue))
    ) {
      triggerPanelChange(date, mergedMode);
    }

    onChange?.(date, info);
  }
};

...

const onInternalSelect = (date: DateType, info: Info = {}) => {
  triggerChange(date, info);
  onSelect?.(date, info);
};
```

**Comment by hbatalhaStch** [comment]
> `selectable` is not a good design API. This mix the business logic into component. It should be like `onChange` update:
> 
> ```tsx
> interface Info {
>   source: 'date' | 'month' | 'year';
> }
> 
> type OnChange<T> = (date: T, info: Info);
> type OnSelect<T> = (date: T, info: Info);
> ```
> 
> Developer can handle this by the logic with:
> 
> ```tsx
> const onMySelect = (date, { source }) => {
>   if (source !== 'date') return;
> };
> ```

Ok, I will do that

> `selectable` is not a good design API. This mix the business logic into component. It should be like `onChange` update:
> 
> ```tsx
> interface Info {
>   source: 'date' | 'month' | 'year';
> }
> 
> type OnChange<T> = (date: T, info: Info);
> type OnSelect<T> = (date: T, info: Info);
> ```
> 
> Developer can handle this by the logic with:
> 
> ```tsx
> const onMySelect = (date, { source }) => {
>   if (source !== 'date') return;
> };
> ```

So this [function](https://github.com/hbatalhaStch/ant-design/blob/ca1dd39dc0d92a8f7be9f5d62cbc4ed51cf08597/components/calendar/generateCalendar.tsx#L204) will be like this:
```ts
const triggerChange = (date: DateType, info: Info) => {
  setMergedValue(date);

  if (!isSameDate(date, mergedValue)) {
    // Trigger when month panel switch month
    if (
      (panelMode === 'date' && !isSameMonth(date, mergedValue)) ||
      (panelMode === 'month' && !isSameYear(date, mergedValue))
    ) {
      triggerPanelChange(date, mergedMode);
    }

    onChange?.(date, info);
  }
};

...

const onInternalSelect = (date: DateType, info: Info = {}) => {
  triggerChange(date, info);
  onSelect?.(date, info);
};
```

File: components/calendar/__tests__/index.test.tsx (hunk 5 above)
Comment (hbatalhaStch) [reply components/calendar/__tests__/index.test.tsx:347]: OK
as [per this comment](https://github.com/ant-design/ant-design/pull/42112#issuecomment-1540013272) we wouldn't even need these tests

**Comment by bombillazo** [comment]
> `selectable` is not a good design API. This mix the business logic into component. It should be like `onChange` update:
> 
> ```tsx
> interface Info {
>   source: 'date' | 'month' | 'year';
> }
> 
> type OnChange<T> = (date: T, info: Info);
> type OnSelect<T> = (date: T, info: Info);
> ```
> 
> Developer can handle this by the logic with:
> 
> ```tsx
> const onMySelect = (date, { source }) => {
>   if (source !== 'date') return;
> };
> ```

It is to have the choice to select what type of date selection triggers callbacks in the calendar, which can be a combination of `date`, `month` or `year` and not limit it to only one.

File: components/calendar/__tests__/index.test.tsx (hunk 6 above)
Comment (hbatalhaStch) [reply components/calendar/__tests__/index.test.tsx:348]: I don't think it is necessary

**Comment by hbatalhaStch** [comment]
@afc163 I have made the changes you suggested, please review
I had some issues using `toHaveBeenLastCalledWith` so I decided to try again later when I have more time available

**Comment by hbatalhaStch** [comment]
@afc163 it says that you merged the branch but the merge commit looks like this:

**Comment by hbatalhaStch** [comment]
@afc163 have you tried to merge this feature? It says that you merged but the merge commit looks like this:

image

it says that "This commit does not belong to any branch on this repository"
image

**Comment by hbatalhaStch** [comment]
> CI failed.

What does this mean?

**Comment by yoyo837** [comment]
image

Try to rebase upstream and fix CI. Thanks.

**Comment by yoyo837** [comment]
`git push -f`

**Comment by hbatalhaStch** [comment]
> `git push -f`

Just did that

**Comment by yoyo837** [comment]
This PR cannot be reopened, please send a new one. Thanks.

**Comment by hbatalhaStch** [comment]
> This PR cannot be reopened, please send a new one. Thanks.

Done. Move to https://github.com/ant-design/ant-design/pull/42407

**Pull request closed without merging**